proposals/0000-splice-imports.rst

```diff
@@ -0,0 +1,353 @@
+   for all dependencies.
+   By distinguishing imported modules whose code is executed only at compile time
+   (which in common cases will be a small fraction of imported modules), we are
+   able to improve this pessimisation.
+2. GHC offers an ``-fno-code`` flag that instructs the compiler to parse and
+   typecheck Haskell modules, but not to generate code. The intent is to offer
+   quick feedback to the user. Any module imports of a module using
+   ``TemplateHaskell`` must be compiled to object code.
+   This is despite the fact that we will not generate object code for the module
+   itself. By distinguishing imported modules whose code is executed only at
+   compile time, we can significantly reduce this unfortunate work, and entirely in many
+   cases.
+3. Projects such as haskell-language-server face similar problems as 2., where they are interested only in the result of type-checking modules, but when ``TemplateHaskell`` is enabled a large
+   number of modules have to be cautiously compiled to bytecode.
+4. By using splice imports we can separate the dependencies into those only needed at build-time (1) and
+   those only needed at runtime (2). We can then link only against those packages needed at runtime.
+5. Currently, when cross-compiling, in modules that use ``TemplateHaskell``, all
+   imported modules must be compiled for both host and target.
+   By distinguishing imported modules used at compile time(i.e. not used at
+   runtime), we can require only those modules to be compiled for the host.
+   Similarly, by distinguishing imported modules used at runtime (i.e. not used at
+   compile time), we can require only those modules to be compiled for the
+   target. It can be very hard or impossible to make some packages available on
+   some cross-compile target platforms, so this change would significantly
+   improve the applicability of ``TemplateHaskell`` in these scenarios.
+
+
+Definitions
+-----------
+
+level
+  Each expression exists at a level. The level is increased by 1 when
+  inside a quote and decreased by 1 inside a splice. Therefore the level of
+  an expression can be calculated as the number of quotes surrounding an expression
+  subtract the number of splices. For example::
+
+    -- foo is at level 0
+    foo = $(let
+      -- bar is at level -1
+      bar = $(let
+        -- baz is at level -2
+        baz = [|
+        -- qux is at level -1
+          qux = [|
+            -- quux is at level 0
+            quux = [|
+              -- quuz is at level 1
+              quuz = 0
+            |]
+          |]
+        |] in baz
+      ) in bar
+    )
+
+
+top-level splice
+  A splice, where the body is at a negative level, an unadorned
+  declaration splice or a quasiquoter.
+
+
+home module
+  A module from the package that is currently being compiled.
+
+
+Proposed Change
+---------------
+
+The ``splice`` modifier indicates that a module's imports are available at compile time.
```

Comment:
This is all in a funny order.  What splice modifier?  Surely you want to put the syntax change first?  Currently it is buried many paragraphs later.

Reply:
This is precisely the description of the `syntax` change. The paragraph states that a new modifier called `splice` is introduced to distinguish between imports availble at compile time or runtime.

Reply:
I have rearranged an reworded this section, I think it's got worse with all the clarifications!